stdlib/LinearAlgebra/src/generic.jl

```diff
@@ -1767,12 +1767,16 @@ promote_leaf_eltypes(x::Union{AbstractArray,Tuple}) = mapreduce(promote_leaf_elt
 # Supports nested arrays; e.g., for `a = [[1,2, [3,4]], 5.0, [6im, [7.0, 8.0]]]`
 # `a ≈ a` is `true`.
 function isapprox(x::AbstractArray, y::AbstractArray;
-    atol::Real=0,
+    atol::Real=zero(promote_type(real(promote_leaf_eltypes(x)),real(promote_leaf_eltypes(y)))),
     rtol::Real=Base.rtoldefault(promote_leaf_eltypes(x),promote_leaf_eltypes(y),atol),
     nans::Bool=false, norm::Function=norm)
     d = norm(x - y)
     if isfinite(d)
-        return d <= max(atol, rtol*max(norm(x), norm(y)))
+        T = promote_type(typeof(atol), typeof(rtol*d))
+        rtol = convert(T, rtol)
+        atol = convert(T, atol)
+        tol = iszero(rtol) ? atol : max(atol, rtol*max(norm(x), norm(y)))
+        return d <= tol
```

Comment:
```suggestion
        iszero(rtol) ? d < atol : d < max(atol, rtol*max(norm(x), norm(y)))
```
The simplest way to avoid type instabilities is often a runtime branch.

Reply:
Much cleaner, agreed - somehow didn't occur to me; thanks!